otherlibs/win32unix/truncate.c

```diff
@@ -0,0 +1,127 @@
+/*                                                                        */
+/*   Copyright 2018 Institut National de Recherche en Informatique et     */
+/*     en Automatique.                                                    */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+#include <sys/types.h>
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/fail.h>
+#include <caml/signals.h>
+#include <caml/io.h>
+#include <caml/osdeps.h>
+#include "unixsupport.h"
+#include <windows.h>
+
+static int win_truncate_handle(HANDLE fh, long long len)
+{
+  LARGE_INTEGER fp;
+  fp.QuadPart = len;
+  if (SetFilePointerEx(fh, fp, NULL, FILE_BEGIN) == 0 ||
+      SetEndOfFile(fh) == 0) {
+    return -1;
+  }
+  return 0;
+}
+
+static int win_ftruncate(HANDLE fh, long long len)
+{
+  HANDLE dupfh, currproc;
+  int ret;
+  if (fh == INVALID_HANDLE_VALUE) {
+    return -1;
+  }
+  currproc = GetCurrentProcess();
+  /* Duplicate the handle, so we are free to modify its file position. */
+  if (DuplicateHandle(currproc, fh, currproc, &dupfh, 0, FALSE,
+                      DUPLICATE_SAME_ACCESS) == 0) {
+     return -1;
+  }
+  ret = win_truncate_handle(dupfh, len);
+  CloseHandle(dupfh);
+  return ret;
+}
+
+static int win_truncate(WCHAR * path, long long len)
+{
+  HANDLE fh;
+  int ret;
+  fh = CreateFile(path, GENERIC_WRITE, 0, NULL,
+                  OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+  if (fh == INVALID_HANDLE_VALUE) {
+    return -1;
+  }
+  ret = win_truncate_handle(fh, len);
+  CloseHandle(fh);
+  return ret;
+}
+
+CAMLprim value unix_truncate(value path, value len)
+{
+  CAMLparam2(path, len);
+  WCHAR * p;
+  int ret;
+  caml_unix_check_path(path, "truncate");
+  p = caml_stat_strdup_to_utf16(String_val(path));
+  caml_enter_blocking_section();
+  ret = win_truncate(p, Long_val(len));
+  caml_leave_blocking_section();
+  caml_stat_free(p);
+  if (ret == -1)
+    uerror("truncate", path);
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value unix_truncate_64(value path, value vlen)
+{
+  CAMLparam2(path, vlen);
+  WCHAR * p;
+  int ret;
+  file_offset len = File_offset_val(vlen);
+  caml_unix_check_path(path, "truncate");
+  p = caml_stat_strdup_to_utf16(String_val(path));
+  caml_enter_blocking_section();
+  ret = win_truncate(p, len);
+  caml_leave_blocking_section();
+  caml_stat_free(p);
+  if (ret == -1)
+    uerror("truncate", path);
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value unix_ftruncate(value fd, value len)
+{
+  CAMLparam1(len);
```

Comment:
```suggestion
  CAMLparam2(fd, len);
```
Equally, these functions don't allocate or do anything with the parameters after the blocking section, so you could just remove the `CAMLparam` completely and just have `return Val_unit;` at the end.

Reply:
Done, got rid of the CAML macros altogether.